src/merkle/mmr/full.rs

```diff
@@ -0,0 +1,232 @@
+    let tree_count = forest.count_ones() as usize;
+    forest * 2 - tree_count
+}
+
+// MMR
+// ===============================================================================================
+
+/// A fully materialized Merkle Mountain Range, with every tree in the forest and all their
+/// elements.
+///
+/// Since this is a full representation of the MMR, elements are never removed and the MMR will
+/// grow roughly `O(n**2)` in number of leaf elements.
+pub struct FullMMR {
+    /// Refer to the `forest` method documentation for details of the semantics of this value.
+    pub(super) forest: usize,
+
+    /// Contains every element of the forest.
+    ///
+    /// The trees are in postorder sequential representation. This representation allows for all
+    /// the elements of every tree in the forest to be stored in the same sequential buffer. It
+    /// also means new elements can be added to the forest, and merging of trees is very cheap with
+    /// no need to copy elements.
+    pub(super) nodes: Vec<Word>,
+}
+
+/// Differentiate the sibling direction on a merkle proof opening.
+///
+/// The MMR peaks have an invariant that their sizes are always distinct, this implies that when a
+/// right child is added to the tree it's merged with the left child right away and a new root is
+/// created. So in a path on the `Left` childs were previously a MMR peak.
+#[derive(PartialEq, Eq, Debug, Clone, Copy)]
+pub enum PathElement<T> {
+    /// Left child and a previous MMR peak.
+    LeftSibling(T),
+    /// Right child, this is never an old MMR peak.
+    RightSibling(T),
+}
+
+impl Default for FullMMR {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl FullMMR {
+    // CONSTRUCTORS
+    // ============================================================================================
+
+    pub fn new() -> FullMMR {
+        FullMMR {
+            forest: 0,
+            nodes: Vec::new(),
+        }
+    }
+
+    // ACCESSORS
+    // ============================================================================================
+
+    /// Returns the MMR forest representation.
+    ///
+    /// The forest value has the following interpretations:
+    /// - its value is the number of elements in the forest
+    /// - bit count corresponds to the number of trees in the forest
+    /// - each true bit position determines the depth of a tree in the forest
+    pub const fn forest(&self) -> usize {
+        self.forest
+    }
+
+    /// Given a leaf position, returns the merkle path to its corresponding peak. If the position
+    /// is greater-or-equal than the tree size None is returned.
+    ///
+    /// Note: The leaf position is the 0-indexed number corresponding to the order the leaves were
+    /// added, this corresponds to the MMR size _prior_ to adding the element. So the 1st alement
+    /// has position 0, the second position 1, and so on.
+    pub fn merkle_path(&self, pos: usize) -> Option<Vec<PathElement<Word>>> {
```

Comment:
I think a slightly different return type would work better for our needs (for MASM, advice provider etc.). Specifically, I'm thinking of something like:
```Rust
pub struct MmrProof {
    peak_idx: usize,
    path: MerklePath,
}
```
Where `MerklePath` is the data structure implemented [here](https://github.com/0xPolygonMiden/crypto/blob/main/src/merkle/path.rs).

This proof would be easy to verify against the `Accumulator` struct.

Reply:
Should we include the relative index of the leaf within the peak.  I believe it would help the client determine the authentication path?

Reply:
Ah yes - some other info would need to be included. I'm not sure if `relative_leaf_idx` is the best option or if something else (e.g., some kind of offset or just total number of leaves) would work better.

Reply:
The reason I suggested `relative_leaf_idx` is because the `MerklePath` verify interface (`pub fn verify(&self, index: u64, node: Word, root: &Word)` expects the index of the node within the merkle tree for verification.  Maybe it's fine to expect the client to compute the index from the total number of leaves but I personally prefer keeping the logic encapsulated here.

Reply:
Just to clarify, I had it implemented using:

```rust
pub enum PathElement<T> {
    LeftSibling(T),
    RightSibling(T),
}
```

The `relative_leaf_idx` and the the encoding above are the same. Namely the `true` bits in `relative_leaf_idx` translates to a `RightSibling`, and the `false` bits translates to a `LeftSibling`. Because I had description, `MerklePath` couldn't be used (it expects `Vec<Word>` and not `Vec<PathElement<Word>>`.

I'm going to remove the `PathElement` in favor of `MerklePath` so that we have more consistent code.

Reply:
Humm, I started this and I want to go back to a point @bobbinth made [earlier](https://github.com/0xPolygonMiden/crypto/pull/67#issuecomment-1437258383):

> > add doesn't return any updates, this is intentional, this will come on a later PR when I implement a sparse version of this structure
>
> I think for now, it may be OK to assume that we just have 2 views of the MMR: (1) full MMR and (2) "accumulator" version with only the peaks.

I think it makes sense to have the sparse version. At least my understanding is that a user will have a Merkle proof `m`, that is complete for a version of the Merkle tree. When the original tree the proof was based on is merged with another, there is a root is added and the proof is no longer complete, so it needs to be updated with the new root. This view of a single proof in the tree is what I meant by sparse implementation, and it corresponds to this return type.

The old version with `Vec<PathElement<Word>>` only needed a new `PathElement<Word>` pushed to the head of the vector with the correct direction `left`/`right`. The version with the position needs equivalent changes.

Reply:
Yes, we will need to have the functionality to update the proofs. I'm thinking it will be something like:
```Rust
proof.update(something).
```
Where `something` is some kind of a descriptor of how the accumulator was updated.

Basically, the idea is that if we have just `Accumulator` view of the MMR, we should be able to produce updates which can be used to update relevant proofs. But I don't know the exact mechanics of this yet.

I still favor using `MerklePath` because it is more consistent with other structures and we'd need to convert to `MerklePath` anyway before sending it to the VM (via advice provider).